control/sysnorm.py

```diff
@@ -0,0 +1,180 @@
+                except Exception as e:
+                    print(f"An error occurred solving the discrete time Lyapunov equation: {e}")
+                    return None
+                
+                # System is stable to reach this point, and P should be positive semi-definite. 
+                # Test next is a precaution in case the Lyapunov equation is ill conditioned.
+                if any(la.eigvals(P) < 0.0):
+                    if print_warning:
+                        print("Warning: There appears to be poles close to the complex unit circle. Norm value may be uncertain.")
+                    return float('inf')
+                else:
+                    norm_value = np.sqrt(np.trace(C@P@C.T + D@D.T))  # Argument in sqrt should be non-negative               
+                    if np.isnan(norm_value):
+                        print("Unknown error. Norm computation resulted in NaN.")           
+                        return None
+                    else:
+                        return norm_value               
+    #
+    # L_infinity-norm computation
+    #
+    elif p == "inf":   
+        def _Hamilton_matrix(gamma):
+            """Constructs Hamiltonian matrix. For internal use."""
+            R = Ip*gamma**2 - D.T@D
+            invR = la.inv(R)
+            return np.block([[A+B@invR@D.T@C, B@invR@B.T], [-C.T@(Ip+D@invR@D.T)@C, -(A+B@invR@D.T@C).T]])    
+    
+        # Discrete time case 
+        # Use inverse bilinear transformation of discrete time system to s-plane if no poles on |z|=1 or z=0.
+        # Allows us to use test for continuous time systems next.
+        if G.isdtime():
+            Ad = A
+            Bd = B
+            Cd = C
+            Dd = D
+            if any(np.isclose(abs(la.eigvals(Ad)), 1.0)):
+                if print_warning:
+                    print("Warning: Poles close to, or on, the complex unit circle. Norm value may be uncertain.")
+                return float('inf')
+            elif any(np.isclose(la.eigvals(Ad), 0.0)):
+                print("L_infinity norm computation for discrete time system with pole(s) at z=0 currently not supported.")            
+                return None
+            
+            # Inverse bilinear transformation
+            In = np.eye(len(Ad))
+            Adinv = la.inv(Ad+In)
+            A = 2*(Ad-In)@Adinv
+            B = 2*Adinv@Bd
+            C = 2*Cd@Adinv
+            D = Dd - Cd@Adinv@Bd
+        
+        # Continus time case 
+        if any(np.isclose(la.eigvals(A).real, 0.0)):
+            if print_warning:
+                print("Warning: Poles close to, or on, imaginary axis. Norm value may be uncertain.")
+            return float('inf')
+    
+        gaml = la.norm(D,ord=2)    # Lower bound
+        gamu = max(1.0, 2.0*gaml)  # Candidate upper bound
+        Ip = np.eye(len(D))    
+         
+        while any(np.isclose(la.eigvals(_Hamilton_matrix(gamu)).real, 0.0)):  # Find actual upper bound
```

Comment:
Have you considered using SLICOT-Routines (already or to be wrapped by Slycot) optimized for certain Hamiltonian matrices task instead of the generic scipy eigenvalue solver? (I didn't check if the SLICOT routines are applicable here)